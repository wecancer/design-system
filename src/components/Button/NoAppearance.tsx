import React from 'react'
import styled from 'styled-components'

const BtnContainer = styled.button`
  appearance: none;
  padding: 0;
  outline: none;
  border: none;
  background: transparent;
`
const ButtonNoAppearance = (props: Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'css'>): React.ReactElement => (
  <BtnContainer type="button" {...props} />
)

export default ButtonNoAppearance
