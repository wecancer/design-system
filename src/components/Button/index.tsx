import React from 'react'
import styled, {css, FlattenSimpleInterpolation} from 'styled-components'
import Loading from '../Loading'

type Color = 'primary' | 'secondary' | 'default'

export const ButtonStyled = styled.button<{colorType: Color}>`
  ${({theme, colorType}) => css`
    height: 40px;
    font-size: 1rem;
    display: inline-block;
    cursor: pointer;
    font-weight: ${theme.font.weights.semiBold};
    padding: 0 1.5rem;
    appearance: none;
    border-radius: 0.5rem;
    outline: none;
    transition: all 250ms ease;

    &:disabled {
      opacity: 0.5;
      cursor: default;
    }

    ${((): FlattenSimpleInterpolation => {
      if (colorType === 'primary') {
        return css`
          color: ${theme.colors.white};
          background-color: ${theme.colors.primary};
          border: 2px solid ${theme.colors.primary};

          &:not(:disabled):hover {
            background-color: ${theme.colors.darkPrimary};
            border-color: ${theme.colors.darkPrimary};
          }

          &:focus {
            box-shadow: 0 0 0.062rem 0.375rem ${theme.colors.focusPrimary};
          }
        `
      }
      if (colorType === 'secondary') {
        return css`
          color: ${theme.colors.primary};
          background-color: ${theme.colors.white};
          border: 2px solid ${theme.colors.primary};

          &:not(:disabled):hover {
            border-color: ${theme.colors.darkPrimary};
          }

          &:focus {
            box-shadow: 0 0 0.062rem 0.375rem ${theme.colors.focusPrimary};
          }
        `
      }
      return css`
        color: ${theme.colors.primary};
        background-color: ${theme.colors.white};
        border: 2px solid ${theme.colors.line};

        &:not(:disabled):hover {
          color: ${theme.colors.darkPrimary};
        }

        &:focus {
          box-shadow: 0 0 0.062rem 0.375rem ${theme.colors.focusLine};
        }
      `
    })()}
  `}
`

export type Props = {
  type?: 'button' | 'reset' | 'submit'
  title?: string
  primary?: boolean
  className?: string
  secondary?: boolean
  isLoading?: boolean
  isDisabled?: boolean
  children: React.ReactNode
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void
}

const Button = ({
  title,
  onClick,
  children,
  className,
  type = 'button',
  primary = false,
  secondary = false,
  isLoading = false,
  isDisabled = false,
}: Props): React.ReactElement => (
  <ButtonStyled
    type={type}
    title={title}
    onClick={onClick}
    disabled={isDisabled}
    className={className}
    colorType={((): Color => {
      if (primary) return 'primary'
      if (secondary) return 'secondary'
      return 'default'
    })()}
  >
    {isLoading ? <Loading /> : children}
  </ButtonStyled>
)

export default Button
