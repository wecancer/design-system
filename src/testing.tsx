import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
