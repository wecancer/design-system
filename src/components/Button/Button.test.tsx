import React from 'react'

import Button from '.'
import {render, fireEvent} from '../../testing'

const btnText = 'Click me!'

describe('<Button />', () => {
  it('should render snapshot correctly', () => {
    const {container} = render(
      <>
        <Button primary>Primary</Button>
        <Button secondary>Secondary</Button>
        <Button isLoading>Loading content!</Button>
        <Button isDisabled>Button disabled</Button>
      </>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the component with text children', () => {
    const {queryByText} = render(<Button>Click me!</Button>)
    expect(queryByText('Click me!')).toBeInstanceOf(HTMLButtonElement)
  })

  it('should add the className in button component', () => {
    const {container} = render(<Button className="customClass">Custom class name</Button>)
    expect(container.querySelector('.customClass')).toBeInstanceOf(HTMLButtonElement)
  })

  it('should receive the isDisabled attribute', () => {
    const {queryByText} = render(<Button isDisabled>{btnText}</Button>)
    expect(queryByText(btnText)).toBeDisabled()
  })

  it('should receive the isLoading attribute', () => {
    const {container} = render(<Button isLoading>Something!</Button>)
    expect(container.querySelector('[data-icon-name="spinnerSolid"]')).toBeDefined()
  })

  it('should receive the onChange attribute', () => {
    const handleClick = jest.fn()
    const {queryByText} = render(<Button onClick={handleClick}>{btnText}</Button>)
    const btnElement = queryByText(btnText)

    if (!btnElement) throw new Error(`Can't find the button with the text "${btnText}"`)

    fireEvent.click(btnElement)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })
})
