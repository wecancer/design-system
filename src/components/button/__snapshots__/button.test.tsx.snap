// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render snapshot correctly 1`] = `
<div>
  <button
    class="sc-gsnTZi cpAPPq"
    type="button"
  >
    Primary
  </button>
  <button
    class="sc-gsnTZi gWbwPf"
    type="button"
  >
    Secondary
  </button>
  <button
    class="sc-gsnTZi CptMN"
    type="button"
  >
    Error
  </button>
  <button
    class="sc-gsnTZi bnDfIJ"
    type="button"
  >
    Success
  </button>
  <button
    class="sc-gsnTZi bQrWvp"
    type="button"
  >
    <div
      class="sc-bczRLJ jPOBKJ"
    >
      <div
        data-icon-name="spinnerSolid"
        style="display: inline-flex; align-items: center; justify-content: center;"
      >
        <svg
          fill="none"
          height="16px"
          viewBox="0 0 20 20"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 1.67A8.33 8.33 0 1018.33 10"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
          />
        </svg>
      </div>
    </div>
  </button>
  <button
    class="sc-gsnTZi bQrWvp"
    disabled=""
    type="button"
  >
    Button disabled
  </button>
</div>
`;
